plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -357,15 +357,48 @@ maybeAddSubst body a s n rhs = do
     hints <- view iiHints
     let hinted = shouldInline hints a n
 
-    preUnconditional <- preInlineUnconditional rhs'
-    if preUnconditional
+    if hinted -- if we've been told specifically, then do it right away
     then extendAndDrop (Done $ dupable rhs')
     else do
-        -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
-        postUnconditional <- postInlineUnconditional rhs'
-        if hinted || postUnconditional
+        -- See Note [Inlining and purity]
+        termIsPure <- checkPurity rhs'
+        let preInlineUnconditional :: InlineM tyname name uni fun a Bool
```

Comment:
I think it would be better to instead have 
```
preInlineUnconditional :: Term -> Bool -> InlineM ...
preInlineUnconditional t isPure = ...
```
Reasons:
1. It keeps the symmetry between the multiple inlining functions: you can just pass them all the purity information
2. This version puts _too much stuff_ in scope in `preInlineUnconditional`: it's now much less clear what stuff is relevant for it. Even the main argument now just comes in from an outer scope!
3. It just keeps the indentation down, which is nice.

If you do that I think you can keep almost exactly the code that was there before, just call `termIsPure` once and pass in the result to the functions.

Reply:
Sure. While I'm here I'll just take them out of `maybeAddSubst`? I'll also rename them. (See my reply to Ziyang's comment)